### PR TITLE
Add size limit to getFileContent (follow up of eclipse-theia/theia#17055)

### PR DESCRIPTION
**Pull request by cdamus**
#### What it does

Adds five commits to address review comments on the PR eclipse-theia/theia#17055 and issues I found in my own testing of it.

1. Implement fixes for critical issues from review: two correctness fixes
   - rewrote the preference description to accurately distinguish that full-file reads check the whole file size while offset/limit reads check only the requested range;
   - replaced .length / 1024 (UTF-16 code units) with new Blob([content]).size / 1024 (actual UTF-8 bytes) for accurate KB measurements on files with multi-byte characters.
3. Implement fixes for significant issues from review:
   - refactoring and robustness: extracted buildFileSizeLimitError / buildSliceSizeLimitError helper methods (DRY)
   - replaced a live preferences.get() call in the tool description with static text
   - added input validation
   - added tests for edge cases
5. Fix brittle mocking issue: cleaned up test mocks to preserve TypeScript type information instead of casting it away, so the compiler can detect API changes that would silently break the mocks.
6. Stream file contents for paginated reads: when the LLM provides offset/limit and the source is an on-disk file, read it via readStream instead of read so that only the requested lines are ever loaded into memory.
7. Fix file-not-found errors on too-large files: when FileService refused to read a file because it exceeded the files.maxFileSizeMB preference, the getFileContent tool was returning a misleading "File not found" error to the LLM rather than the pagination suggestion. Fixed by bypassing the files.maxFileSizeMB cap in the streaming path (since streaming never loads the full file into memory) and by properly recognising FILE_TOO_LARGE / FILE_EXCEEDS_MEMORY_LIMIT errors in catch blocks to return the correct guidance.

The latter two commits actually are key for the original motivation of the upstream PR: the original scenario required the LLM to analyze the format of very large files in order to compose scripts intended to process the data that they contain. This would never have worked without a streaming solution because these files are too large for the FileService to read in one shot anyways, so the original PR is effectively moot without streaming support.

#### How to test

According to the upstream PR description. Also, run the new tests in this PR and try asking an agent to analyze the structure of a file of 300 MB size or greater.

#### Follow-ups

None identified for this PR; they would all be upstream.

#### Breaking changes

- [ ] This PR introduces breaking changes and requires careful review. If yes, the breaking changes section in the [changelog](https://github.com/eclipse-theia/theia/blob/master/CHANGELOG.md) has been updated.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)
- [x] User-facing text is internationalized using the `nls` service (for details, please see the [Internationalization/Localization section](https://github.com/theia-ide/theia/blob/master/doc/coding-guidelines.md#internationalizationlocalization) in the Coding Guidelines)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/ai-ide/src/browser/workspace-functions.ts [hunk 1]
```diff
@@ -345,15 +369,151 @@ export class FileContentFunction implements ToolProvider {
...
+                    // Full-file reads (startLine=0, no limit) should advise the caller to use
+                    // offset/limit for pagination.  Explicit slice requests should advise using
+                    // a smaller limit instead.
+                    const sizeError = (startLine === 0 && limit === undefined)
+                        ? this.buildFileSizeLimitError(resultSizeKB, maxSizeKB)
+                        : this.buildSliceSizeLimitError(resultSizeKB, maxSizeKB);
+                    resolve(sizeError);
+                    return;
+                }
+                const header = `[Lines ${startLine + 1}\u2013${startLine + sliceLines.length} of ${lineIndex} total. Use offset and limit to read other ranges.]`;
```
Comment (eneufeld) [inline packages/ai-ide/src/browser/workspace-functions.ts]: we always return the header even if limit is undefined and startLine is set to 0. the other case in getFileContent doesn't add this header in this case

File: packages/ai-ide/src/browser/workspace-functions.ts
Comment (eneufeld) [inline packages/ai-ide/src/browser/workspace-functions.ts]: I think we need to extend the logic here, if limit/offset are provided we might want to see them too

File: packages/ai-ide/src/browser/workspace-functions.ts [hunk 2]
```diff
@@ -320,16 +334,26 @@ export class FileContentFunction implements ToolProvider {
...
+    protected readonly preferences: PreferenceService;
+
+    private parseArg(arg_string: string): { file: string; offset?: number; limit?: number } {
         const result = JSON.parse(arg_string);
-        return result.file;
+        return { file: result.file, offset: result.offset, limit: result.limit };
     }
 
-    private async getFileContent(file: string, cancellationToken?: CancellationToken): Promise<string> {
+    private async getFileContent(file: string, cancellationToken?: CancellationToken, offset?: number, limit?: number): Promise<string> {
```
Comment (eneufeld) [inline packages/ai-ide/src/browser/workspace-functions.ts:346]: the method is quite complex, can we split it into something like this:

```
const isEditorOpen = openEditorValue !== undefined;
const isPaginated  = offset !== undefined || limit !== undefined;

if (isEditorOpen) {
    return this.handleEditorContent(openEditorValue!, offset, limit, maxSizeKB);
} else if (isPaginated) {
    return this.readStreamedSlice(targetUri, offset ?? 0, limit, maxSizeKB);
} else {
    return this.handleFullDiskRead(targetUri, maxSizeKB);
}
```

that would also solve the issue that the catch does not use the `buildFileSizeLimitError` method

**Comment by cdamus** [comment]
Thanks @eneufeld for the new feedback. This should all be addressed by commit 05a5e82b5a0311d9b6a9a6780509281604475b39, including a change to the strategy for counting kilobytes that works better in the unit tests (i.e. is node-compatible). With these changes, Architect is actually able to more efficiently read through a parts of a large file to understand its format and come much more quickly to a plan for a Python script to parse and analyze its content.

File: packages/ai-ide/src/browser/workspace-functions.ts
Comment (cdamus) [reply packages/ai-ide/src/browser/workspace-functions.ts]: I opted just to set hasMore for the ellipsis in the tool-call short label as a cue to the user that it was something different to a simple slurp of the whole file as I think details of the offset and limit wouldn't be especially meaningful.

**Comment by eneufeld** [comment]
@cdamus could you squash and I will merge it then

**Comment by cdamus** [comment]
Thanks @eneufeld cf. commit 555c818cb61663f338758082e9d0d2129b88d39f
